File: find_three_largest_numbers/find_three_largest_numbers.py
def find_three_largest_numbers(array):
  three_largest = [None, None, None]
  for num in array:
    update_largest(three_largest, num)
  return three_largest

def update_largest(three_largest, num):
  if three_largest[2] is None or num > three_largest[2]:
    shift_and_update(three_largest, num, 2)
  elif three_largest[1] is None or num > three_largest[1]:
    shift_and_update(three_largest, num, 1)
  elif three_largest[0] is None or num > three_largest[0]:
    shift_and_update(three_largest, num, 0)

def shift_and_update(array, num, idx):
  for i in range(idx + 1):
    if i == idx:
      array[i] = num
    else:
      array[i] = array[i + 1]

print(find_three_largest_numbers([10, 5, 9, 10, 12]));

File: find_three_largest_numbers/find_three_largest_numbers.go
package main

import (
	"fmt"
)

func findThreeLargestNumbers(array []int) [3]int {
	threeLargest := [3]int{-1 << 63, -1 << 63, -1 << 63} // int64 
	for _, num := range array {
		updateLargest(&threeLargest, num)
	}
	return threeLargest
}

func updateLargest(threeLargest *[3]int, num int) {
	if num > threeLargest[2] {
		shiftAndUpdate(threeLargest, num, 2)
	} else if num > threeLargest[1] {
		shiftAndUpdate(threeLargest, num, 1)
	} else if num > threeLargest[0] {
		shiftAndUpdate(threeLargest, num, 0)
	}
}

func shiftAndUpdate(array *[3]int, num int, idx int) {
	for i := 0; i <= idx; i++ {
		if i == idx {
			array[i] = num
		} else {
			array[i] = array[i+1]
		}
	}
}

func main() {
	result := findThreeLargestNumbers([]int{10, 5, 9, 10, 12})
	fmt.Println(result[0], result[1], result[2])
}
